src/target/adiv5.c

```diff
@@ -766,18 +764,27 @@ adiv5_access_port_s *adiv5_new_ap(adiv5_debug_port_s *dp, uint8_t apsel)
 			/* Otherwise note this is a 64-bit AP and read the high part */
 			ap.flags |= ADIV5_AP_FLAGS_64BIT;
 			ap.base |= (uint64_t)adiv5_ap_read(&ap, ADIV5_AP_BASE_HIGH) << 32U;
-		}
-		/* Check the Debug Base Address register for not-present. See ADIv5 Specification C2.6.1 */
-		if (base_flags == (ADIV5_AP_BASE_FORMAT_ADIV5 | ADIV5_AP_BASE_PRESENT_NO_ENTRY) ||
-			(!(ap.flags & ADIV5_AP_FLAGS_64BIT) && (uint32_t)ap.base == ADIV5_AP_BASE_NOT_PRESENT)) {
-			/*
-			 * Debug Base Address not present in this MEM-AP
-			 * No debug entries... useless AP
-			 * AP0 on STM32MP157C reads 0x00000002
-			 */
+			/* Check the Debug Base Address register for not-present (legacy format). See ADIv5 Specification C2.6.1 */
+		} else if ((uint32_t)ap.base == ADIV5_AP_BASE_NOT_PRESENT) {
 			DEBUG_INFO(" -> Not Present\n");
 			return NULL;
 		}
+		/* Make sure we only pay attention to the base address, not the presence and format bits */
+		ap.base &= ADIV5_AP_BASE_BASEADDR;
+		/* Check for ADIv5 no-entry flag. */
+		if (base_flags == (ADIV5_AP_BASE_FORMAT_ADIV5 | ADIV5_AP_BASE_PRESENT_NO_ENTRY)) {
+			/*
+			* Debug Base Address not present in this MEM-AP
+			* No debug entries... useless AP
+			* AP0 on STM32MP157C reads 0x00000002
+			* Unless it's a TI MSPM0, which reads present flag as no entry while reporting a ROM table
+			*/
+			if (!(dp->target_designer_code == JEP106_MANUFACTURER_TEXAS && ap.base != 0U)) {
+				DEBUG_INFO(" -> Not Present\n");
+				return NULL;
+			}
+		}
+
```

Comment:
We're not especially a fan that this introduces a repeated return (this code is fine fwiw, we'd just like not to have the condition return repeated), perhaps this suggestion will work (also noting we don't find the inserted extra new line particularly helpful, so that's dropped in the suggestion)?

```suggestion
		/* Check the Debug Base Address register for not-present. See ADIv5 Specification C2.6.1 */
		if (base_flags == (ADIV5_AP_BASE_FORMAT_ADIV5 | ADIV5_AP_BASE_PRESENT_NO_ENTRY) ||
			(!(ap.flags & ADIV5_AP_FLAGS_64BIT) && (uint32_t)ap.base == ADIV5_AP_BASE_NOT_PRESENT)) {
			/*
			 * Debug Base Address not present in this MEM-AP
			 * No debug entries... useless AP
			 * AP0 on STM32MP157C reads 0x00000002
			 * 
			 * NB: MSPM0 parts erroneously set BASE.P = 0 despite there being
			 * valid debug components on AP0, so we have to have an exception
			 * for this part family.
			 */
			if (dp->target_designer_code != JEP106_MANUFACTURER_TEXAS || (ap.base & ADIV5_AP_BASE_BASEADDR) == 0U) {
				DEBUG_INFO(" -> Not Present\n");
				return NULL;
			}
		}
		/* Make sure we only pay attention to the base address, not the presence and format bits */
		ap.base &= ADIV5_AP_BASE_BASEADDR;
```

Reply:
Now it's only almost correct. This probably won't ever happen in real life as legacy is legacy, but with the proposed change if the BASE is in legacy and equal to -1 (aka not present) and it's TI then testing a masked ap.base for non-null will succeed and such AP will be deemed as present. Culling out this case (legacy not present) early and forgetting about it seems easier albeit with an additional return.

One goto could solve the additional return problem though ;)

In my opinion the proposed version makes the code harder on the logic like multiple tests for the same flags (64-bit) and masking ap.base twice, but if the goal is to optimize for the lowest return statements possible, then this is the way to go. Then because legacy not-present and ADIv5 no-entry checks are lumped together, would need to add check against 64bit and ADIV5_AP_BASE_NOT_PRESENT in the if that checks againt TI designer code. I'm not too fond of it.

Reply:
We'd really like to avoid a goto because that turns the codegen to absolute slop; Given this very specific case and that we know the expected base value is `0xf0000002U`, and you indeed pointing out we got the logic wrong there for legacy.. that would make the inner guard:
```c
if (dp->target_designer_code != JEP106_MANUFACTURER_TEXAS || ap.base != 0xf0000002U) {
```

That helps to key directly to this specific exception without breaking the check for legacy not-present, or over-complicating the checks etc. Does that work for you?

Reply:
Right, it does create a few more lines vertical code, but it simplifies the conditions which I would argue increases the maintainability, not decreases it. I support the concern regarding forgetting to update the text if modifying the messages, but aside from increasing executable size if not updated in sync that is a minor issue. Speaking of size, your proposed is 26 bytes extra compared to my version (either goto or string duplication). The difference is only 4 bytes if TI check is removed.

Anyway, while I still think my version is simpler (and happens to be more compact) you're the maintainer so it's your decision. I'm fine making the code as you proposed and let's move on.